kong/plugins/ai-proxy/schema.lua

```diff
@@ -0,0 +1,125 @@
+  fields = {
+    { max_tokens = { type = "integer", description = "Defines the max_tokens, if using chat or completion models.", required = false, default = 256 } },
+    { temperature = { type = "number", description = "Defines the matching temperature, if using chat or completion models.", required = false, between = { 0.0, 5.0 }, default = 1.0 } },
+    { top_p = { type = "number", description = "Defines the top-p probability mass, if supported.", required = false, between = { 0, 1 }, default = 1.0 } },
+    { top_k = { type = "integer", description = "Defines the top-k most likely tokens, if supported.", required = false,between = { 0, 500 }, default = 0 } },
+    { anthropic_version = { type = "string", description = "Defines the schema/API version, if using Anthropic provider.", required = false } },
+    { azure_instance = { type = "string", description = "Instance name for Azure OpenAI hosted models.", required = false } },
+    { llama2_format = { type = "string", description = "If using llama2 provider, select the upstream message format.", required = false, one_of = { "raw", "openai"} } },
+    { upstream_url = typedefs.url { description = "Manually specify or override the full URL to the AI operation endpoints, when calling (self-)hosted models, or for running via a private endpoint.", required = false } },
+  }
+}
+
+local model_schema = {
+  type = "record",
+  required = true,
+  fields = {
+    { provider = { type = "string", description = "AI provider request format - Kong translates requests to and from the specified backend compatible formats.", required = true, one_of = { "openai", "azure", "anthropic", "cohere", "mistral", "llama2", "preserve" }, default = "openai" } },
+    { name = { type = "string", description = "Model name to execute.", required = true, } },
+    { options = model_options_schema },
+  }
+}
+
+local logging_schema = {
+  type = "record",
+  required = true,
+  fields = {
+    { log_statistics = { type = "boolean", description = "If enabled and supported by the driver, will add model usage and token metrics into the Kong log plugin(s) output.", required = true, default = true } },
+    { log_payloads = { type = "boolean", description = "If enabled, will log the request and response body into the Kong log plugin(s) output.", required = true, default = false } },
+  }
+}
+
+return {
+  name = "ai-proxy",
+  fields = {
+    { protocols = typedefs.protocols_http },
+    { consumer = typedefs.no_consumer },
+    { config = {
+      type = "record",
+      fields = {
+        { route_type = { type = "string", description = "The model's operation implementation, for this provider.", required = true, one_of = { "llm/v1/chat", "llm/v1/completions" } } },
+        { auth = auth_schema },
+        { model = model_schema },
+        { logging = logging_schema },
+      }
+    }}
+  },
+  entity_checks = {
+    { custom_entity_check = {
+      field_sources = { "config" },
+      fn = function(entity)
```

Comment:
I think much of what's in this custom function can be expressed in a declarative manner in our entity_check definitions.

There are some good working examples of this sort of thing on the [upstream](https://github.com/Kong/kong/blob/7f93a9292be1bbf413666fc304cc889fba5ba58b/kong/db/schema/entities/upstreams.lua#L208-L295) and [service](https://github.com/Kong/kong/blob/7f93a9292be1bbf413666fc304cc889fba5ba58b/kong/db/schema/entities/services.lua#L52-L73) entities. The plugin config schemas (`kong/plugins/*/schema.lua`) and custom entity schemas `(kong/plugins/*/daos.lua`) contain examples as well.

Please try to convert whatever you can to declarative entity checks. I know the schema system can be a little mysterious at first, so don't hesitate to reach out if you get stuck.

Reply:
I tried this before, but it just doesn't work for plugins. I think it broke at some point.

Even a simple entity_check fails:

```
/usr/local/share/lua/5.1/kong/db/schema/init.lua:1243: attempt to index a nil value

stack traceback:
	/usr/local/share/lua/5.1/kong/db/schema/init.lua:1243: in function 'run_entity_check'
	/usr/local/share/lua/5.1/kong/db/schema/init.lua:1360: in function 'run_checks'
	/usr/local/share/lua/5.1/kong/db/schema/init.lua:1394: in function 'run_entity_checks'
	/usr/local/share/lua/5.1/kong/db/schema/init.lua:2025: in function 'validate_insert'
	./spec/helpers.lua:602: in function 'validate'
```

I gave up and programmed myself.

Reply:
I didn't know you could use DAO entity options in plugins.

I've changed this now.